framework/src/iotbus/iotbus_gpio.c

```diff
@@ -400,88 +412,142 @@ int iotbus_gpio_unregister_cb(iotbus_gpio_context_h dev)
 	return IOTBUS_ERROR_NONE;
 }
 
-#ifdef CONFIG_IOTDEV
-int iotbus_gpio_set_interrupt(iotbus_gpio_context_h dev, iotbus_int_type_e int_type, iotbus_gpio_cb cb)
+static pthread_addr_t gpio_intr_thread(pthread_addr_t arg)
 {
-	int fd, ret, edge;
+	iotbus_gpio_context_h dev = (iotbus_gpio_context_h)arg;
 	struct _iotbus_gpio_s *handle;
-	struct iotbus_int_info_s info = { 0, };
+	
+	siginfo_t info;
+	sigset_t sig_set;
+
+	sigemptyset(&sig_set);
+	sigaddset(&sig_set, SIG_IOTBUS_GPIO_RISING);
+	sigaddset(&sig_set, SIG_IOTBUS_GPIO_FALLING);
+
+	pthread_sigmask(SIG_BLOCK, &sig_set, NULL);
 
 	handle = (struct _iotbus_gpio_s *)dev->handle;
-	handle->cb = cb;
+	if (handle->cb == NULL) {
+		return NULL;
+	}
 
-	info.handle = dev;
-	info.pin_type = IOTBUS_GPIO;
-	info.int_type = int_type;
-	info.pid = iotapi_get_pid();
+	while (1) {
+		sigwaitinfo(&sig_set, &info);
+		// Call callback function.
+		(handle->cb)(dev);
+	}
 	
-	fd = open(IOTBUS_SIGPATH, O_NONBLOCK);
-	if (fd < 0) {
-		ret = get_errno();
-		ibdbg("Open iotsig fail[%d, %d]\n", fd, ret);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+	pthread_sigmask(SIG_UNBLOCK, &sig_set, NULL);
+	pthread_exit(NULL);
+
+	return NULL;
+}
+
+int iotbus_gpio_set_interrupt(iotbus_gpio_context_h dev, iotbus_gpio_edge_e int_type, iotbus_gpio_cb cb, uint8_t priority)
+{
+	int ret = -1;
+	struct _iotbus_gpio_s *handle;
+	FAR struct gpio_notify_s notify;
+
+	if (!dev || !dev->handle || int_type == IOTBUS_GPIO_EDGE_NONE) {
+		return IOTBUS_ERROR_INVALID_PARAMETER;
 	}
 
-	if (int_type == IOTBUS_GPIO_FALLING) {
-		edge = GPIO_EDGE_FALLING;
-	} else if (int_type == IOTBUS_GPIO_EDGE_RISING) {
-		edge = GPIO_EDGE_RISING;
-	} else {
-		edge = GPIO_EDGE_NONE;
+	handle = (struct _iotbus_gpio_s *)dev->handle;
+
+	if (handle->cb != NULL || handle->pid > 0) {
+		return IOTBUS_ERROR_INVALID_PARAMETER;
 	}
 
-	ret = ioctl(fd, IOTBUS_INTR_REGISTER, (unsigned long)&info);
-	if (ret < 0) {
-		ibdbg("Fail to register %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+	switch (int_type) {
+	case IOTBUS_GPIO_EDGE_BOTH:
+		return IOTBUS_ERROR_NOT_SUPPORTED;
+	case IOTBUS_GPIO_EDGE_RISING:
+		notify.gn_rising  = true;
+		notify.gn_falling = false;
+		notify.gn_signo = SIG_IOTBUS_GPIO_RISING;
+		break;
+	case IOTBUS_GPIO_EDGE_FALLING:
+		notify.gn_rising  = false;
+		notify.gn_falling = true;
+		notify.gn_signo = SIG_IOTBUS_GPIO_FALLING;
+		break;
+	default:
+		return IOTBUS_ERROR_INVALID_PARAMETER;
 	}
-	ret = ioctl(handle->fd, GPIOIOC_SET_INTERRUPT, edge);
-	if (ret < 0) {
-		ibdbg("Fail to set %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+
+	struct intr_attr val = { priority, IOTBUS_GPIO, getpid() };
+
+	handle->cb = cb;
+	handle->pid = create_intr_pthread(gpio_intr_thread, (void *)dev, &val);
+	if (handle->pid <= 0) {
+		handle->cb = NULL;
+		handle->pid = 0;
+		return IOTBUS_ERROR_UNKNOWN;
```

Comment:
IOTBUS_ERROR_UNKNOWN is not acceptable, you already know why it failed. (pid < 0) You have to use another one, this is very important reason.